File: swift/Sources/benchmarks/Matmul4t.swift
import Dispatch
import Foundation

class Matmul4T: BenchmarkProtocol {
  public var n: Int64 = 0
  public var resultVal: UInt32 = 0
  private let lock = NSLock()

  init() {
    n = configValue("n") ?? 0
  }

  func getNumThreads() -> Int {
    return 4
  }

  private func matgen(_ n: Int) -> [[Double]] {
    let tmp = 1.0 / Double(n) / Double(n)
    var a = [[Double]](repeating: [Double](repeating: 0, count: n), count: n)

    for i in 0..<n {
      for j in 0..<n {
        a[i][j] = tmp * Double(i - j) * Double(i + j)
      }
    }
    return a
  }

  private func matmulParallel(_ a: [[Double]], _ b: [[Double]]) -> [[Double]] {
    let size = a.count
    let numThreads = getNumThreads()

    var bT = [[Double]](repeating: [Double](repeating: 0, count: size), count: size)
    for i in 0..<size {
      for j in 0..<size {
        bT[j][i] = b[i][j]
      }
    }

    var c = [[Double]](repeating: [Double](repeating: 0, count: size), count: size)

    DispatchQueue.concurrentPerform(iterations: numThreads) { threadId in
      let startRow = threadId
      for i in stride(from: startRow, to: size, by: numThreads) {
        let ai = a[i]
        var row = [Double](repeating: 0.0, count: size)

        for j in 0..<size {
          var sum = 0.0
          let bTj = bT[j]

          for k in 0..<size {
            sum += ai[k] * bTj[k]
          }

          row[j] = sum
        }

        lock.lock()
        c[i] = row
        lock.unlock()
      }
    }

    return c
  }

  func run(iterationId: Int) {
    let a = matgen(Int(n))
    let b = matgen(Int(n))
    let c = matmulParallel(a, b)
    let center = c[Int(n) / 2][Int(n) / 2]
    resultVal &+= Helper.checksumF64(center)
  }

  var checksum: UInt32 {
    return resultVal
  }

  func prepare() {}

  func name() -> String {
    return "Matmul::T4"
  }
}

File: kotlin/src/main/kotlin/benchmarks/Matmul4T.kt
package benchmarks
import Benchmark

import kotlinx.coroutines.*
import java.util.concurrent.Executors

class Matmul4T : Benchmark() {
    private var n: Int = 0
    private var resultValue: Long = 0L
    
    init {
        n = iterations
    }
    
    private fun matgen(n: Int): Array<DoubleArray> {
        val tmp = 1.0 / n / n
        val a = Array(n) { DoubleArray(n) }
        
        for (i in 0 until n) {
            for (j in 0 until n) {
                a[i][j] = tmp * (i - j) * (i + j)
            }
        }
        return a
    }
    
    private fun matmulParallel(a: Array<DoubleArray>, b: Array<DoubleArray>): Array<DoubleArray> {
        val size = a.size
        
        // Транспонируем b (последовательно)
        val bT = Array(size) { DoubleArray(size) }
        for (i in 0 until size) {
            for (j in 0 until size) {
                bT[j][i] = b[i][j]
            }
        }
        
        // Умножение матриц (параллельно)
        val c = Array(size) { DoubleArray(size) }
        
        // Используем фиксированный пул потоков
        val numThreads = 4
        val executor = Executors.newFixedThreadPool(numThreads)
        val scope = CoroutineScope(executor.asCoroutineDispatcher())
        
        runBlocking {
            val jobs = mutableListOf<Job>()
            val rowsPerThread = (size + 3) / numThreads
            
            for (threadId in 0 until numThreads) {
                val job = scope.launch {
                    val startRow = threadId * rowsPerThread
                    val endRow = minOf(startRow + rowsPerThread, size)
                    
                    for (i in startRow until endRow) {
                        val ai = a[i]
                        val ci = c[i]
                        
                        for (j in 0 until size) {
                            var sum = 0.0
                            val bTj = bT[j]
                            
                            for (k in 0 until size) {
                                sum += ai[k] * bTj[k]
                            }
                            
                            ci[j] = sum
                        }
                    }
                }
                jobs.add(job)
            }
            
            // Ждем завершения всех корутин
            jobs.joinAll()
        }
        
        executor.shutdown()
        return c
    }
    
    override fun run() {
        val a = matgen(n)
        val b = matgen(n)
        val c = matmulParallel(a, b)
        val center = c[n shr 1][n shr 1]
        resultValue = Helper.checksumF64(center).toLong()
    }
    
    override val result: Long
        get() = resultValue
}
